### ISSUE TITLE
Some historic messages are undecryptable, but are loaded much later

### ISSUE THREAD

**Issue by richvdh** [matrix-org/matrix-rust-sdk#4429]
Sometimes, often when you scroll a long way up your timeline, you encounter occasional unable-to-decrypt messages. (Empirically, it seems like the further back you go, the more UTD messages you find.)

Often, if you come back later on, some of those UTD messages are decrypted.

When we look at the logs, we see:
 * Decryption error
 * Key retrieved from backup
 * But no attempt to re-decrypt (until much later)

In other words, this sounds much the same as #3786, but it's intermittent.

**Comment by richvdh** [comment]
One suggestion was that the homeserver is overloaded with GET /room_keys requests, and starts rejecting them -- but this doesn't seem to be the case. As per the description, the keys appear to be successfully retrieved from backup; they just aren't used.

Another suggestion is that the keys are being retrieved from backup during the notification-sync-poll, and the main "thread" doesn't receive a memo that the key has been retrieved. This seems inconsistent given it affects quite old messages, though.

**Comment by richvdh** [comment]
> Hopefully this will be fixed by [#3872](https://github.com/matrix-org/matrix-rust-sdk/issues/3872)

We've recently seen a couple of rageshakes with similar symptoms, so it seems like it has not been completely fixed.

@poljar is going to have a think about adding some logging to try and diagnose what's going wrong.
